### DISCUSSION TITLE
Multiplexor matrixbutton in out

### DISCUSSION THREAD

**Discussion by pmjcontroller1** [tttapa/Control-Surface#906] [Q&A]
good afternoon, I need to make two matrices and I need to use multiplexers, the idea is to use a multiplexer for the columns and another for the rows.  the question is if the multiplexers could be used as input and as output?

**Comment by tttapa** [comment]
No, multiplexers can only be used as inputs.  
You could implement a custom button matrix class that supports multiplexers for the outputs, but the general ExtIO multiplexer classes cannot support outputs, because only one output can be active at any given time, and this relies on the address pins not changing after changing the pin to output mode. which cannot be guaranteed in general.

See the implementation of ButtonMatrix, it's quite straightforward:

https://github.com/tttapa/Control-Surface/blob/39182bbab79487236c11579db3acfa336813bf54/src/AH/Hardware/ButtonMatrix.ipp#L15-L40

**Reply by pmjcontroller1** [reply]
Hi Pieter,
I'm thinking to use the MCP23017 to control a keyboard matrix. Is it possible with the current library?
And in case that I need two MCP in the i2c bus, how can affect to the performance?
Of course, the mcp will be attached to Arduino ProMicro.
Thanks!!!!

**Reply by tttapa** [reply]
In theory, Control Surface's MCP23017 implementation supports all operations required to read a keyboard matrix. I've never tried it in practice, though, so be sure to start with a small-scale test on a breadboard. You might also need external pull-up resistors, the MCP23017's built-in pull-ups are not particularly strong.
